README
# Advanced Combat Framework

The Advanced Combat Framework is a [Garry's Mod][] addon which creates a balanced, realistic and fun to play combat damage system, simulating projectiles, armor and explosions, as well as engines and gearboxes.

[Garry's Mod]: <http://garrysmod.com/>


changelog.txt
r500
--For an extended version history, please visit the github changelog.  https://github.com/nrlulz/ACF/commits/master

*r555
-Added ground gas turbines
-Added automatic transmissions
-Added 76mm SA
-Added client convar ACF_MobilityRopeLinks, setting 0 will skip rope creation on your links
-Added ACF E2/SF/EA2 function "entity:acfHitClip( vector hitpos )"; returns 1 if hitpos is on clipped part of entity
-Added ACF E2/SF/EA2 function "acfHoldGear" and "acfShiftPointScale" for automatic gearboxes
-Added a hook for ACF_KEShove
-Added legality checks for SOLID_NONE on ACF entities
-Added Sestze's MASK_SHOT passthrough (lets 30mm and under pass through things like chain-link fences)
-Added power to weight ratio output for armor properties tool reload function
-Added a few small optimizations and behind the scenes improvements
-Updated readme with install instructions, and a mention of Knight Icy's weapon sound pack
-Updated ACF Sound Tool to support a future update to ACF Missiles
-Updated engine stats for electrics and turbines to show the true powerband
-Fixed operation order error on visclip passthrough
-Fixed engines ocassionally refusing to use fuel
-Fixed lua error when trying to create ammo for nonexistent weapon types
-Fixed possible lua error on KE shove
-Reduced volume of ammo refill
-Reduced FL minimum spread from 1.0 to 0.25
-Damaged gearboxes lose efficiency
-More relaxed weldparenting rules on mobility (allows parenting chains, requires weld to last [physical] parent)
-Allowed diffs, double diffs, transfer cases, and clutches to be parented without welds
-Buffed smoke radius 25%
-Buffed smoke filler deploy rate, WP filler lifetime
-Buffed petrol health
-Buffed electric health massively
-Nerfed aero turbine throttle response
-Nerfed transaxial aero turbines to 80% of inline equivalent
-Nerfed large aero turbine power 8.5%, weight 10%
-Buffed small inline aero turbine power 15%
-Buffed med inline aero turbine power 10%
-Buffed small and med electric power 20%
-Buffed large electric power 40%
-Nerfed 0.9L I2 and 1.0L I4 special weight by 15kg
-Slight nerf to 5.2L V6
-Slight buff to 7.8L V8

*r558
-Updated SF functions to new standard, courtesy of EpicG
-Fixed armor tool reload not working at all
-Hopefully fixed armor tool reload sometimes not getting entire contraption
-Fixed transaxial GG turbines
-Fixed issues with auto gearbox UI
-Fixed too many digits on auto gearbox overlay
-Fixed issue with double shifting on gearup/geardown inputs

*r560
-github pls

*r561
-Fixed change in r555 that broke missiles

*r562
-Fixed change in r555 that broke standalone clutch
-Clamped added time from round extraction to half of load time, when unloading gun

*r563
-Made it easier for plugins to modify crate text.

*r564
-Fixed sound tool shoot effect.
-Added extension support to sound tool.
-Added error notifications to sound tool.

*r565
-Fixed sbox_acf_restrictinfo support with EA2 component.

*r569
-Make acf_bulleteffect use SetEntity() instead of SetMagnitude()

*r573
-Minor optimization of fuel tanks, ammo crates, gearboxes, and engines
-Minor optimization of ballistics courtesy of Red
-Removed ACFs SetParent override, no longer needed
-Fixed gearboxes not working after updating settings
-Fixed auto-gearbox shift point generator UI
-Increased random dispersion of ground ricochets
-Fixed prop ricochet angles, added some random dispersion
-Partial fix for hitting yourself when firing at high speed
-Partial fix for bullets ghosting through props
-Improved ballistics backtracing, increased backtrace length
-Fix for the ocassional bullet exploding when reentering world courtesy of red
-Added a few bugfixes courtesy of Steve
--Fixed issue with world penetration not properly reducing KE
--Fixed bullets getting culled early near source map limits
--Fixed issue with parented guns failing to give bullets inherited velocity
-Made muzzleflash smoke and missile smoke less blinding
-Fuel tank capacity calculation improved, better representation of model
--Capacity increased ~50%, but still 1/8th realistic for balance
--Refuel rate doubled
-Added e2 function acfDragDriv()
-Reverted to old dragdiv 40, increasing ballistic drag
--Minimal effect on cannons, reduces effective range of light weapons like MG
-clamped short round firerate bonus to 19%
-Smoke rounds updated
--Added timed fuse to smoke ammo (min 0.5s), set to 0 to disable timed fuse
--Allowed M and HW to fire smoke rounds
--Smoke effect can handle up to 10,000g WP and 10,000g SM filler (up from 200g)
--WP duration reduced 27% (from 29 seconds to 21 seconds for full length 40mm SL WP)
--Added relevant info to ammocrate overlay
-Increased max guns to 16 to help offset racks/flares using gun limit
-Added rate of fire and magazine info to guns on acf menu tool
-Reduced 40mm SL weight from 25 kg to 5 kg
-Increased 40mm GL reload from 2.5s to 7.5s
-Increased 37mm C round length from 45 cm to 48 cm
-Adjusted MGs
--7.62mm RoF increased from 477 to 600 r/m
--7.62mm given 250 round magazine with 6 second reload
--12.7mm RoF increased from 460 to 600 r/m
--12.7mm given 150 round magazine with 6 second reload
--14.5mm RoF increased from 346 to 480 r/m
--14.5mm given 90 round magazine with 5 second reload
--20mm given 50 round magazine with 4 second reload
-Adjusted HMGs courtesy of Red
--Spread increased from 0.32 to 0.4
--20mm 160kg, RoF 675 r/m, 12 rounds, 8s reload, 40mm pen @ 850 m/s
--30mm 480kg, RoF 434 r/m, 10 rounds, 10s reload, 56mm pen @ 850 m/s
--40mm 740kg, RoF 380 r/m, 8 rounds, 13s reload, 69mm pen @ 850 m/s
-Adjusted SAs
--Increased SA spread from 0.08 to 0.12
--Decreased 25mm RoF from 214 to 180 r/m
--Increased 37mm weight from 480 kg to 540 kg

lua/autorun/acf_globals.lua
ACF = {}
ACF.AmmoTypes = {}
ACF.MenuFunc = {}
ACF.AmmoBlacklist = {}
ACF.Version = 588 -- REMEMBER TO CHANGE THIS FOR GODS SAKE, OMFG!!!!!!! -wrex   Update the changelog too! -Ferv
ACF.CurrentVersion = 0 -- just defining a variable, do not change

ACF.Year = 1945

ACF.Threshold = 264.7	--Health Divisor (don't forget to update cvar function down below)
ACF.PartialPenPenalty = 5 --Exponent for the damage penalty for partial penetration
ACF.PenAreaMod = 0.85
ACF.KinFudgeFactor = 2.1	--True kinetic would be 2, over that it's speed biaised, below it's mass biaised
ACF.KEtoRHA = 0.25		--Empirical conversion from (kinetic energy in KJ)/(Aera in Cm2) to RHA penetration
ACF.GroundtoRHA = 0.15		--How much mm of steel is a mm of ground worth (Real soil is about 0.15)
ACF.KEtoSpall = 1
ACF.AmmoMod = 0.6		-- Ammo modifier. 1 is 1x the amount of ammo
ACF.ArmorMod = 1
ACF.Spalling = 0
ACF.GunfireEnabled = true
ACF.MeshCalcEnabled = false

ACF.HEPower = 8000		--HE Filler power per KG in KJ
ACF.HEDensity = 1.65	--HE Filler density (That's TNT density)
ACF.HEFrag = 1500		--Mean fragment number for equal weight TNT and casing
ACF.HEBlastPen = 0.4	--Blast penetration exponent based of HE power
ACF.HEFeatherExp = 0.5 	--exponent applied to HE dist/maxdist feathering, <1 will increasingly bias toward max damage until sharp falloff at outer edge of range

ACF.HEATMVScale = 0.74	--Filler KE to HEAT slug KE conversion expotential
ACF.HEATMulAmmo = 16.5 		--HEAT slug damage multiplier; 13.2x roughly equal to AP damage
ACF.HEATMulFuel = 8.25		--needs less multiplier, much less health than ammo
ACF.HEATMulEngine = 8.25	--likewise

ACF.DragDiv = 40		--Drag fudge factor
ACF.VelScale = 1		--Scale factor for the shell velocities in the game world
-- local PhysEnv = physenv.GetPerformanceSettings()
ACF.PhysMaxVel = 4000
ACF.SmokeWind = 5 + math.random()*35 --affects the ability of smoke to be used for screening effect

ACF.PBase = 1050		--1KG of propellant produces this much KE at the muzzle, in kj
ACF.PScale = 1	--Gun Propellant power expotential
ACF.MVScale = 0.5  --Propellant to MV convertion expotential
ACF.PDensity = 1.6	--Gun propellant density (Real powders go from 0.7 to 1.6, i'm using higher densities to simulate case bottlenecking)

ACF.TorqueBoost = 1.25 --torque multiplier from using fuel
ACF.FuelRate = 5  --multiplier for fuel usage, 1.0 is approx real world
ACF.ElecRate = 1.5 --multiplier for electrics
ACF.TankVolumeMul = 0.5 -- multiplier for fuel tank capacity, 1.0 is approx real world

ACF.FuelDensity = { --kg/liter
	Diesel = 0.832,  
	Petrol = 0.745,
	Electric = 3.89 -- li-ion
}

ACF.Efficiency = { --how efficient various engine types are, higher is worse
	GenericPetrol = 0.304, --kg per kw hr
	GenericDiesel = 0.243, --up to 0.274
	Turbine = 0.375, -- previously 0.231
	Wankel = 0.335,
	Radial = 0.4, -- 0.38 to 0.53
	Electric = 0.85 --percent efficiency converting chemical kw into mechanical kw
}

ACF.TorqueScale = { --how fast damage drops torque, lower loses more % torque
	GenericPetrol = 0.25,
	GenericDiesel = 0.35,
	Turbine = 0.2,
	Wankel = 0.2,
	Radial = 0.25,
	Electric = 0.5
}

ACF.EngineHPMult = { --health multiplier for engines
	GenericPetrol = 0.2,
	GenericDiesel = 0.5,
	Turbine = 0.125,
	Wankel = 0.125,
	Radial = 0.2,
	Electric = 0.75
}

ACF.LiIonED = 0.458 -- li-ion energy density: kw hours / liter
ACF.CuIToLiter = 0.0163871 -- cubic inches to liters

ACF.RefillDistance = 300 --Distance in which ammo crate starts refilling.
ACF.RefillSpeed = 700 -- (ACF.RefillSpeed / RoundMass) / Distance 

ACF.DebrisScale = 20 -- Ignore debris that is less than this bounding radius.
ACF.SpreadScale = 4		-- The maximum amount that damage can decrease a gun's accuracy.  Default 4x
ACF.GunInaccuracyScale = 1 -- A multiplier for gun accuracy.
ACF.GunInaccuracyBias = 2  -- Higher numbers make shots more likely to be inaccurate.  Choose between 0.5 to 4. Default is 2 (unbiased).

ACF.EnableDefaultDP = false -- Enable the inbuilt damage protection system.


if file.Exists("acf/shared/acf_userconfig.lua", "LUA") then
	include("acf/shared/acf_userconfig.lua")
end


CreateConVar('sbox_max_acf_gun', 16)
CreateConVar('sbox_max_acf_smokelauncher', 10)
CreateConVar('sbox_max_acf_ammo', 32)
CreateConVar('sbox_max_acf_misc', 32)
CreateConVar('acf_meshvalue', 1)
CreateConVar("sbox_acf_restrictinfo", 1) -- 0=any, 1=owned

AddCSLuaFile()
AddCSLuaFile( "acf/client/cl_acfballistics.lua" )
AddCSLuaFile( "acf/client/cl_acfmenu_gui.lua" )
AddCSLuaFile( "acf/client/cl_acfrender.lua" )

if SERVER and ACF.EnableDefaultDP then
	AddCSLuaFile( "acf/client/cl_acfpermission.lua" )
	AddCSLuaFile( "acf/client/gui/cl_acfsetpermission.lua" )
end

if SERVER then

	util.AddNetworkString( "ACF_KilledByACF" )
	util.AddNetworkString( "ACF_RenderDamage" )
	util.AddNetworkString( "ACF_Notify" )

	include("acf/server/sv_acfbase.lua")
	include("acf/server/sv_acfdamage.lua")
	include("acf/server/sv_acfballistics.lua")
	
	if ACF.EnableDefaultDP then
		include("acf/server/sv_acfpermission.lua")
	end

elseif CLIENT then

	include("acf/client/cl_acfballistics.lua")
	include("acf/client/cl_acfrender.lua")
	
	if ACF.EnableDefaultDP then
		include("acf/client/cl_acfpermission.lua")
		include("acf/client/gui/cl_acfsetpermission.lua")
	end
	
	killicon.Add( "acf_AC", "HUD/killicons/acf_AC", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_AL", "HUD/killicons/acf_AL", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_C", "HUD/killicons/acf_C", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_GL", "HUD/killicons/acf_GL", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_HMG", "HUD/killicons/acf_HMG", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_HW", "HUD/killicons/acf_HW", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_MG", "HUD/killicons/acf_MG", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_MO", "HUD/killicons/acf_MO", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_RAC", "HUD/killicons/acf_RAC", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_SA", "HUD/killicons/acf_SA", Color( 200, 200, 48, 255 ) )
	killicon.Add( "acf_ammo", "HUD/killicons/acf_ammo", Color( 200, 200, 48, 255 ) )
	
	CreateConVar("acf_cl_particlemul", 1)
	CreateClientConVar("ACF_MobilityRopeLinks", "1", true, true)
	
	-- Cache results so we don't need to do expensive filesystem checks every time
	local IsValidCache = {}

	-- Returns whether or not a sound actually exists, fixes client timeout issues
	function IsValidSound( path )
		if IsValidCache[path] == nil then 
			IsValidCache[path] = file.Exists( string.format( "sound/%s", tostring( path ) ), "GAME" ) and true or false
		end
		return IsValidCache[path]
	end
	
end

include("acf/shared/rounds/roundap.lua")
include("acf/shared/rounds/roundaphe.lua")
include("acf/shared/rounds/roundhe.lua")
include("acf/shared/rounds/roundheat.lua")
include("acf/shared/rounds/roundfl.lua")
include("acf/shared/rounds/roundhp.lua")
include("acf/shared/rounds/roundsmoke.lua")
include("acf/shared/rounds/roundrefill.lua")
include("acf/shared/rounds/roundfunctions.lua")

include("acf/shared/acfloader.lua")
include("acf/shared/acfcratelist.lua")
--include("acf/shared/acfmissilelist.lua")

ACF.Weapons = list.Get("ACFEnts")
	
ACF.Classes = list.Get("ACFClasses")

ACF.RoundTypes = list.Get("ACFRoundTypes")

ACF.IdRounds = list.Get("ACFIdRounds")	--Lookup tables so i can get rounds classes from clientside with just an integer

game.AddParticles("particles/acf_muzzleflashes.pcf")
game.AddParticles("particles/explosion1.pcf")
game.AddParticles("particles/rocket_motor.pcf")

game.AddDecal("GunShot1", "decals/METAL/shot5")

-- Add the ACF tool category
if CLIENT then

	ACF.CustomToolCategory = CreateClientConVar( "acf_tool_category", 0, true, false );

	if( ACF.CustomToolCategory:GetBool() ) then

		language.Add( "spawnmenu.tools.acf", "ACF" );

		-- We use this hook so that the ACF category is always at the top
		hook.Add( "AddToolMenuTabs", "CreateACFCategory", function()

			spawnmenu.AddToolCategory( "Main", "ACF", "#spawnmenu.tools.acf" );

		end );

	end

end

timer.Simple( 0, function()
	for Class,Table in pairs(ACF.Classes["GunClass"]) do
		PrecacheParticleSystem(Table["muzzleflash"])
	end
end)

--Stupid workaround red added to precache timescaling.
hook.Add( "Think", "Update ACF Internal Clock", function()
	ACF.CurTime = CurTime()
    ACF.SysTime = SysTime()
end )

-- changes here will be automatically reflected in the armor properties tool
function ACF_CalcArmor( Area, Ductility, Mass )
	
	return ( Mass * 1000 / Area / 0.78 ) / ( 1 + Ductility ) ^ 0.5 * ACF.ArmorMod
	
end

function ACF_MuzzleVelocity( Propellant, Mass, Caliber )

	local PEnergy = ACF.PBase * ((1+Propellant)^ACF.PScale-1)
	local Speed = ((PEnergy*2000/Mass)^ACF.MVScale)
	local Final = Speed -- - Speed * math.Clamp(Speed/2000,0,0.5)

	return Final
end

function ACF_Kinetic( Speed , Mass, LimitVel )
	
	LimitVel = LimitVel or 99999
	Speed = Speed/39.37
	
	local Energy = {}
		Energy.Kinetic = ((Mass) * ((Speed)^2))/2000 --Energy in KiloJoules
		Energy.Momentum = (Speed * Mass)
		
		local KE = (Mass * (Speed^ACF.KinFudgeFactor))/2000 + Energy.Momentum
		Energy.Penetration = math.max( KE - (math.max(Speed-LimitVel,0)^2)/(LimitVel*5) * (KE/200)^0.95 , KE*0.1 )
		--Energy.Penetration = math.max( KE - (math.max(Speed-LimitVel,0)^2)/(LimitVel*5) * (KE/200)^0.95 , KE*0.1 )
		--Energy.Penetration = math.max(Energy.Momentum^ACF.KinFudgeFactor - math.max(Speed-LimitVel,0)/(LimitVel*5) * Energy.Momentum , Energy.Momentum*0.1)
	
	return Energy
end

-- returns last parent in chain, which has physics
function ACF_GetPhysicalParent( obj )
	if not IsValid(obj) then return nil end
	
	--check for fresh cached parent
	if obj.acfphysparent and ACF.CurTime < obj.acfphysstale then
		return obj.acfphysparent
	end
	
	local Parent = obj
	
	while Parent:GetParent():IsValid() do
		Parent = Parent:GetParent()
	end
	
	--update cached parent
	obj.acfphysparent = Parent
	obj.acfphysstale = ACF.CurTime + 10 --when cached parent is considered stale and needs updating
	
	return Parent
end

-- Global Ratio Setting Function
function ACF_CalcMassRatio( obj, pwr )
	if not IsValid(obj) then return end
	local Mass = 0
	local PhysMass = 0
	local power = 0
	local fuel = 0
	
	-- find the physical parent highest up the chain
	local Parent = ACF_GetPhysicalParent(obj)
	
	-- get the shit that is physically attached to the vehicle
	local PhysEnts = ACF_GetAllPhysicalConstraints( Parent )
	
	-- add any parented but not constrained props you sneaky bastards
	local AllEnts = table.Copy( PhysEnts )
	for k, v in pairs( AllEnts ) do
		
		table.Merge( AllEnts, ACF_GetAllChildren( v ) )
	
	end
	
	for k, v in pairs( AllEnts ) do
		
		if IsValid( v ) then
		
			if v:GetClass() == "acf_engine" then
				power = power + (v.peakkw * 1.34)
				fuel = v.RequiresFuel and 2 or fuel
			elseif v:GetClass() == "acf_fueltank" then
				fuel = math.max(fuel,1)
			end
			
			local phys = v:GetPhysicsObject()
			if IsValid( phys ) then		
			
				Mass = Mass + phys:GetMass()
				
				if PhysEnts[ v ] then
					PhysMass = PhysMass + phys:GetMass()
				end
				
			end
		
		end
		
	end
	
	for k, v in pairs( AllEnts ) do
		v.acfphystotal = PhysMass
		v.acftotal = Mass
		v.acflastupdatemass = ACF.CurTime
	end
	
	if pwr then return { Power = power, Fuel = fuel } end
end

-- Cvars for recoil/he push
CreateConVar("acf_hepush", 1)
CreateConVar("acf_recoilpush", 1)

-- New healthmod/armormod/ammomod cvars
CreateConVar("acf_healthmod", 1)
CreateConVar("acf_armormod", 1)
CreateConVar("acf_ammomod", 1)
CreateConVar("acf_spalling", 0)
CreateConVar("acf_gunfire", 1)

function ACF_CVarChangeCallback(CVar, Prev, New)
	if( CVar == "acf_healthmod" ) then
		ACF.Threshold = 264.7 / math.max(New, 0.01)
		print ("Health Mod changed to a factor of " .. New)
	elseif( CVar == "acf_armormod" ) then
		ACF.ArmorMod = 1 * math.max(New, 0)
		print ("Armor Mod changed to a factor of " .. New)
	elseif( CVar == "acf_ammomod" ) then
		ACF.AmmoMod = 1 * math.max(New, 0.01)
		print ("Ammo Mod changed to a factor of " .. New)
	elseif( CVar == "acf_spalling" ) then
		ACF.Spalling = math.floor(math.Clamp(New, 0, 1))
		local text = "off"
		if(ACF.Spalling > 0) then
			text = "on"
		end
		print ("ACF Spalling is now " .. text)
	elseif( CVar == "acf_gunfire" ) then
		ACF.GunfireEnabled = tobool( New )
		local text = "disabled"
		if ACF.GunfireEnabled then 
			text = "enabled" 
		end
		print ("ACF Gunfire has been " .. text)
	end
end

if SERVER then
	function ACF_SendNotify( ply, success, msg )
		net.Start( "ACF_Notify" )
			net.WriteBit( success )
			net.WriteString( msg or "" )
		net.Send( ply )
	end
else
	local function ACF_Notify()
		local Type = NOTIFY_ERROR
		if tobool( net.ReadBit() ) then Type = NOTIFY_GENERIC end
		
		GAMEMODE:AddNotify( net.ReadString(), Type, 7 )
	end
	net.Receive( "ACF_Notify", ACF_Notify )
end

function ACF_UpdateChecking( )
	http.Fetch("https://github.com/nrlulz/ACF",function(contents,size)
		local rev = tonumber(string.match( contents, "%s*(%d+)\n%s*</span>\n%s*commits" )) or 0 --"history\"></span>\n%s*(%d+)\n%s*</span>"
		if rev and ACF.Version >= rev then
			print("[ACF] ACF Is Up To Date, Latest Version: "..rev)
		elseif !rev then
			print("[ACF] No Internet Connection Detected! ACF Update Check Failed")
		else
			print("[ACF] A newer version of ACF is available! Version: "..rev..", You have Version: "..ACF.Version)
			if CLIENT then chat.AddText( Color( 255, 0, 0 ), "A newer version of ACF is available!" ) end
		end
		ACF.CurrentVersion = rev
		
	end, function() end)
end
ACF_UpdateChecking( )

local function OnInitialSpawn( ply )
	local Table = {}
	for k,v in pairs( ents.GetAll() ) do
		if v.ACF and v.ACF.PrHealth then
			table.insert(Table,{ID = v:EntIndex(), Health = v.ACF.Health, v.ACF.MaxHealth})
		end
	end
	if Table ~= {} then
		net.Start("ACF_RenderDamage")
			net.WriteTable(Table)
		net.Send(ply)
	end
end

hook.Add( "PlayerInitialSpawn", "renderdamage", OnInitialSpawn )

cvars.AddChangeCallback("acf_healthmod", ACF_CVarChangeCallback)
cvars.AddChangeCallback("acf_armormod", ACF_CVarChangeCallback)
cvars.AddChangeCallback("acf_ammomod", ACF_CVarChangeCallback)
cvars.AddChangeCallback("acf_spalling", ACF_CVarChangeCallback)
cvars.AddChangeCallback("acf_gunfire", ACF_CVarChangeCallback)

-- smoke-wind cvar handling
if SERVER then
	local function msgtoconsole(hud, msg)
			print(msg)
	end

	util.AddNetworkString("acf_smokewind")
	concommand.Add( "acf_smokewind", function(ply, cmd, args, str)
			local validply = IsValid(ply)
			local printmsg = validply and function(hud, msg) ply:PrintMessage(hud, msg) end or msgtoconsole
			
			if not args[1] then printmsg(HUD_PRINTCONSOLE,
					"Set the wind intensity upon all smoke munitions." ..
					"\n   This affects the ability of smoke to be used for screening effect." ..
					"\n   Example; acf_smokewind 300")
					return false
			end
			
			if validply and not ply:IsAdmin() then
					printmsg(HUD_PRINTCONSOLE, "You can't use this because you are not an admin.")
					return false
					
			else
					local wind = tonumber(args[1])

					if not wind then
							printmsg(HUD_PRINTCONSOLE, "Command unsuccessful: that wind value could not be interpreted as a number!")
							return false
					end
					
					ACF.SmokeWind = wind
					
					net.Start("acf_smokewind")
							net.WriteFloat(wind)
					net.Broadcast()
					
					printmsg(HUD_PRINTCONSOLE, "Command SUCCESSFUL: set smoke-wind to " .. wind .. "!")
					return true        
			end
	end)

	local function sendSmokeWind(ply)
			net.Start("acf_smokewind")
					net.WriteFloat(ACF.SmokeWind)
			net.Send(ply)
	end
	hook.Add( "PlayerInitialSpawn", "ACF_SendSmokeWind", sendSmokeWind )
else
	local function recvSmokeWind(len)
		ACF.SmokeWind = net.ReadFloat()
	end
	net.Receive("acf_smokewind", recvSmokeWind)
end

/*
ONE HUGE HACK to get good killicons.
*/
-- disabling this for now because it was breaking killicons completely and i don't want to deal with it right now
/*
if SERVER then
	
	hook.Add("PlayerDeath", "ACF_PlayerDeath",function( victim, inflictor, attacker )
		if inflictor:GetClass() == "acf_ammo" then
			net.Start("ACF_KilledByACF")
				net.WriteString( victim:Nick()..";ammo;"..attacker:Nick() )
			net.Broadcast()
		end
		if inflictor:GetClass() == "acf_gun" then
			net.Start("ACF_KilledByACF")
				net.WriteString( victim:Nick()..";"..inflictor.Class..";"..attacker:Nick() )
			net.Broadcast()
		end
	end)
end


if CLIENT then
	
	net.Receive("ACF_KilledByACF", function()
		local Table = string.Explode(";", net.ReadString())
		local victim, gun, attacker = Table[1], Table[2], Table[3]
		
		if attacker == "worldspawn" then attacker = "" end
		GAMEMODE:AddDeathNotice( attacker, -1, "acf_"..gun, victim, 1001 )
	end)
	
	if not ACF.replacedPlayerKilled then
		timer.Create("ACF_replacePlayerKilled", 1, 0, function()
			local Hooks = usermessage.GetTable()
			if Hooks["PlayerKilled"] then
				local ACF_PlayerKilled = Hooks["PlayerKilled"].Function
				ACF.replacedPlayerKilled = true
				Hooks["PlayerKilled"].Function = function(msg)
					local victim     = msg:ReadEntity()
					if ( !IsValid( victim ) ) then return end
					local inflictor    = msg:ReadString()
					local attacker     = msg:ReadString()
					if inflictor != "acf_gun" and inflictor != "acf_ammo" then
						ACF_PlayerKilled(msg)
					end
				end
				timer.Destroy("ACF_replacePlayerKilled")
				Msg("[ACF] Replaced PlayerKilled\n")
			end
		end)
	end
	if not ACF.replacedPlayerKilledByPlayer then
		timer.Create("ACF_replacePlayerKilledByPlayer", 1, 0, function()
			local Hooks = usermessage.GetTable()
			if Hooks["PlayerKilledByPlayer"] then
				local ACF_PlayerKilledByPlayer = Hooks["PlayerKilledByPlayer"].Function
				ACF.replacedPlayerKilledByPlayer = true
				Hooks["PlayerKilledByPlayer"].Function = function(msg)
					local victim     = msg:ReadEntity()
					local inflictor    = msg:ReadString()
					local attacker     = msg:ReadEntity()

					if ( !IsValid( attacker ) ) then return end
					if ( !IsValid( victim ) ) then return end
					if inflictor != "acf_gun" and inflictor != "acf_ammo" then
						ACF_PlayerKilledByPlayer(msg)
					end
				end
				timer.Destroy("ACF_replacePlayerKilledByPlayer")
				Msg("[ACF] Replaced PlayerKilledByPlayer\n")
			end
		end)
	end
end
*/
